text/0323-array-functions.md

```diff
@@ -0,0 +1,130 @@
+- Start Date: 2018-03-25
+- RFC PR: https://github.com/emberjs/rfcs/pull/323
+- Ember Issue: (leave this empty)
+
+# Array functions
+
+## Summary
+
+This RFC proposes a way to work with arrays in ember without requiring that we extend the global `Array` prototype or manually wrap arrays with `Ember.A(array)`.
+
+Specifically, for each method on the `MutableArray` mixin we will provide a similarly named function that takes an array as its first argument followed by the arguments to the method.
+
+For example, instead of calling `array.pushObject(item)` you can call `pushObject(array, item)`. For a full list see the expandible list in [Detailed design](#detailed-design).
+
+## Motivation
+
+Currently it is not a great experience to write an app with `Array` prototype extensions disabled. You end up needing to wrap arrays everywhere with `Ember.A(array)` which loops through each method on the `NativeArray` mixin and assigns it as an enumerable own property on the array.
+
+The are several problems here:
+
+- It's annoying to defensively write `Ember.A()` around every array
+- It's wasteful to looping through all the array methods and assign them all. Especially if you're only going to use one of them
+- Iterating over an array with `for..in` or `Object.keys` will not behave as generally expected because the methods are assigned enumerably (for performance reasons)
+
+All of these problems go away when using array functions instead of array methods.
+
+It's also worth noting that when you using fastboot you are required to have `Array` prototype extensions disabled currently and thus forced to deal with these issues.
+
+## Detailed design
+
+For each method in the `MutableArray` mixin we will export a similarly named function from `@ember/array`.
+
+For example, the method `array.objectAt(index)` will map to the function `objectAt(array, index)` with implementation
+
+```js
+function objectAt(array, index) {
+  if (Array.isArray(array)) {
+    return array[index];
+  } else {
+    return array.objectAt(index);
+  }
+}
+```
```

Comment:
so the new functions are delegating to the old methods? wouldn't that mean we're duplicating the implementation (once in the new functions for native arrays and once in the old methods for extended prototype arrays)?

Reply:
> so the new functions are delegating to the old methods?

Yes, for non-Array objects we must call the method.

> wouldn't that mean we're duplicating the implementation (once in the new functions for native arrays and once in the old methods for extended prototype arrays)?

In the existing code, each implementer of `MutableArray` must supply an `objectAt` and a `replace` method. I wouldn't say that this could is "duplicated" but rather that each class has a "specialized" version of the method. This proposal is simply to shuffle around the existing code.

For example, if NativeArray and ArrayProxy were implemented as

```js
export let NativeArray = Mixin.create(MutableArray, {
  replace() { replaceInNativeArray(this, ...arguments); }
});

export let ArrayProxy = EmberObject.create(MutableArray, {
  replace() { /* something complex */ }
});
```

then we would just be adding the function

```js
function replace(array, ...args) {
  if (Array.isArray(array)) {
    replaceInNativeArray(array, ...args);
  } else {
    array.replace(...args);
  }
}
```

As you can see the duplication of logic is minimal here because the logic is shared in the `replaceInNativeArray` function. (In fact, that's how the code is arranged on `master` today: https://github.com/emberjs/ember.js/blob/master/packages/ember-metal/lib/array.js#L18-L24).

This issue is targeted at making these functions public.